jwst/badpix_selfcal/badpix_selfcal.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import annotations
+import numpy as np
+import jwst.datamodels as dm
+from jwst.outlier_detection.outlier_detection_ifu import medfilt
+from stdatamodels.jwst.datamodels.dqflags import pixel
+
+
+def badpix_selfcal(minimg: np.ndarray,
+                   flagfrac: float = 0.001,
+                   kernel_size: int = 15,
+                   dispaxis=None) -> np.ndarray:
+    """
+    Flag residual artifacts as bad pixels in the DQ array of a JWST exposure
+
+    Parameters
+    ----------
+    minimg: np.ndarray
+        Selfcal data of shape (x, y), i.e., after some operation has
+        already been taken to combine multiple exposures,
+        typically a MIN operation.
+    flagfrac: float
+        Fraction of pixels to flag on each of low and high end
+    kernel_size: int
+        Size of kernel for median filter
+    dispaxis: int
+        Dispersion axis, either 1 or 2. If None, a two-dimensional
+        median filter is applied.
+
+    Returns
+    -------
+    flagged_indices: np.ndarray
+        Indices of the flagged pixels,
+        shaped like output from np.where
+    """
+    if flagfrac < 0 or flagfrac >= 0.5:
+        raise ValueError("flagfrac must be between 0 and 0.5. \
+                        Note this fraction will be flagged on both high and low ends.")
+    if dispaxis not in [1, 2, None]:
+        raise ValueError("dispaxis must be either 1 or 2, or None.")
+
+    # Determine outliers using median filter
+    if dispaxis is None:
+        kern_size = (kernel_size, kernel_size)
+    elif dispaxis == 2:
+        # check that this is the right way around!
+        kern_size = (kernel_size, 1)
+    elif dispaxis == 1:
+        kern_size = (1, kernel_size)
+
+    smoothed = medfilt(minimg, kern_size)
+    minimg_hpf = minimg - smoothed
+
+    # Flag outliers using percentile cutoff
+    flag_low, flag_high = np.nanpercentile(minimg_hpf, [flagfrac * 100, (1 - flagfrac) * 100])
+    bad = (minimg_hpf > flag_high) | (minimg_hpf < flag_low)
+    flagged_indices = np.where(bad)
+
+    return flagged_indices
+
+
+def apply_flags(input_model: dm.IFUImageModel, flagged_indices: np.ndarray) -> dm.IFUImageModel:
+    """
+    Apply the flagged indices to the input model. Sets the flagged pixels to NaN
+    and the DQ flag to DO_NOT_USE + OTHER_BAD_PIXEL
+
+    Parameters
+    ----------
+    input_model: dm.IFUImageModel
+        Input science data to be corrected
+    flagged_indices: np.ndarray
+        Indices of the flagged pixels,
+        shaped like output from np.where
+
+    Returns
+    -------
+    output_model: dm.IFUImageModel
+        Flagged data model
+    """
+
+    input_model.dq[flagged_indices] = pixel["DO_NOT_USE"] + pixel["OTHER_BAD_PIXEL"]
```

Comment:
Here, it's probably advisable to use a bit-wise or ('|') instead of just setting directly to these two flags. That way, you can preserve any additional informational flags that pixel might have already had.

Reply:
I updated it to be a `+=` operation instead.  My understanding of the flagging scheme is that this will preserve all the information because all sums of DQ flags are unique.  Is that right?  It wasn't immediately obvious to me why using `or` is correct here.

Reply:
It should be |, not +.  You need a bitwise-or because a pixel might already have one of the flags you're trying to set.  If it already has DO_NOT_USE, then += DO_NOT_USE would set it to 2 instead of 1, which is a completely different flag.